Вы аналитик российской авиакомпании F9, выполняющей внутренние пассажирские перевозки. Важно понять предпочтения пользователей, покупающих билеты на разные направления.

Вам предстоит изучить базу данных и проанализировать спрос пассажиров на рейсы в города, где проходят крупнейшие культурные фестивали.


<div class="alert alert-info">
Привет! Меня зовут Никита Мишин и я буду твоим ревьюером по этому проекты.
Для простоты предлагаю общение на 'ты'. Буду предполагать, что ты не против:) 
Если более предпочтительно обращение на 'Вы', пиши, не стесняйся.
Также если будут возникать вопросы, аналогично, пиши:)

Предлагаю работать в известном тебе итеративном формате.
Итерация состоит в моей проверке твоего решения. 
После решения могут остаться какие-то недочеты, которые я попрошу тебя устранить, ты их исправляешь и я проверяю твои решения.
    Оставленные мною комментарии могут быть разного вида:
   
    - зеленый: элегантные решения, которые тебе стоит запомнит и в дальнейшем взять на вооружение:) 
    
    - желтый: сигнал о том, что есть некритичная вещь(не всегда ошибка), что нужно точно поправить в следующей работе, даже желательно в этой (полезно, в первую очередь, для тебя:) ).Также это рекомендации на будущее    

    - красный: недочет, который нужно исправить в этой работе, для того, чтобы она была принята
    
    - синий: полезная информация, доп ресурсы, "вопросы на подумать"

Также попрошу не удалять мои комментарии:) <a class="tocSkip">
</div>

**Аналитика средствами Python**

У вас есть файлы, в которых содержатся результаты запросов из предыдущих заданий.

**query_1.csv** — результат первого запроса. В нём содержится такая информация:

- *model* — **модель самолета;
- *flights_amount* — количество рейсов для каждой модели самолетов *model* в сентябре 2018 года.

**query_3.csv** — результат третьего запроса. В нём содержится такая информация:

- *city* — город;
- *average_flights* — среднее количество рейсов, прибывающих в город (*city*) за день в сентябре 2018 года.

 Для этих двух наборов данных нужно:

- импортировать файлы;
- изучить данные в них;
- проверить типы данных на корректность;
- выбрать топ-10 городов по количеству рейсов;
- построить графики: модели самолётов и количество рейсов, города и количество рейсов, топ-10 городов и количество рейсов;
- сделать выводы по каждому из графиков, пояснить результат.


# Аналитика

In [1]:
!pip install cufflinks plotly -U

Defaulting to user installation because normal site-packages is not writeable
Requirement already up-to-date: cufflinks in /home/jovyan/.local/lib/python3.7/site-packages (0.17.3)
Requirement already up-to-date: plotly in /home/jovyan/.local/lib/python3.7/site-packages (4.14.3)


In [2]:
import os

import numpy as np
import pandas as pd

import seaborn as sns
import matplotlib.pyplot as plt


import math
from scipy import stats as st

import requests

# импортируем модули plotly для интерактивной визуализации
import plotly.graph_objs as go
import cufflinks as cf
import plotly.express as px
from plotly import __version__
from plotly.offline import download_plotlyjs, init_notebook_mode, plot, iplot 
init_notebook_mode(connected=True)
cf.go_offline()

PATH = '/datasets/'
SEED = 21


<div class="alert alert-success">
<h1>Комментарий ревьюера <a class="tocSkip"></a></h1>
Верное решение о расположении всех импортов в начале тетрадки. Так пользователь сразу поймет, какие нужны библиотеки
</div>

## Исследование данных

### Импорт файлов

In [3]:
query_1 =( pd.read_csv(os.path.join(PATH, 'query_1.csv'))
          .sort_values(by = 'flights_amount',ascending = False)
          .reset_index(drop= True))
query_3 =( pd.read_csv(os.path.join(PATH, 'query_3.csv'))
          .sort_values(by= 'average_flights', ascending =False)
          .reset_index(drop=True))

### Проверка данных

In [4]:
print('Модели самолётов и количество рейсов')
print('------------------------------------------')
print(query_1.head(10))
print('------------------------------------------')
print(query_1.info())
print('------------------------------------------')
print(query_1.describe())

Модели самолётов и количество рейсов
------------------------------------------
                 model  flights_amount
0   Cessna 208 Caravan            4557
1   Bombardier CRJ-200            4446
2  Sukhoi SuperJet-100            4185
3      Airbus A321-200             960
4       Boeing 737-300             630
5      Airbus A319-100             607
6       Boeing 767-300             600
7       Boeing 777-300             300
------------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8 entries, 0 to 7
Data columns (total 2 columns):
model             8 non-null object
flights_amount    8 non-null int64
dtypes: int64(1), object(1)
memory usage: 256.0+ bytes
None
------------------------------------------
       flights_amount
count        8.000000
mean      2035.625000
std       1965.203947
min        300.000000
25%        605.250000
50%        795.000000
75%       4250.250000
max       4557.000000


In [5]:
print('Города и количество рейсов, среднее в день')
print('------------------------------------------')
print(query_3.head(10))
print('------------------------------------------')
print(query_3.info())
print('------------------------------------------')
print(query_3.describe())

Города и количество рейсов, среднее в день
------------------------------------------
              city  average_flights
0           Москва       129.774194
1  Санкт-Петербург        31.161290
2      Новосибирск        17.322581
3       Красноярск        11.580645
4     Екатеринбург        11.322581
5   Ростов-на-Дону        10.193548
6            Пермь        10.129032
7           Брянск        10.000000
8             Сочи         9.612903
9        Ульяновск         9.580645
------------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101 entries, 0 to 100
Data columns (total 2 columns):
city               101 non-null object
average_flights    101 non-null float64
dtypes: float64(1), object(1)
memory usage: 1.7+ KB
None
------------------------------------------
       average_flights
count       101.000000
mean          5.494189
std          13.119790
min           1.000000
25%           2.000000
50%           3.000000
75%           5.870968
max     


<div class="alert alert-warning">
<h1>Комментарий ревьюера <a class="tocSkip"></a></h1>
Дублируется функциональность для разных датасетов. Стоило завести функцию с данной общей функциональностью
</div>

### проверка на наличие NaN значений

In [6]:
nan_1 = query_1.isna().sum()

nan_2 = query_3.isna().sum()

print('NaN flights amount:\n', nan_1, '\n\n NaN average_flights:\n', nan_2)

NaN flights amount:
 model             0
flights_amount    0
dtype: int64 

 NaN average_flights:
 city               0
average_flights    0
dtype: int64


### Проверка на наличие дубликатов

In [7]:
duplicate_1 = query_1.duplicated().sum()

duplicate_2 = query_3.duplicated().sum()

print( 'Duplicated flights amount:', duplicate_1,'\n Duplicated average flights:', duplicate_2)

Duplicated flights amount: 0 
 Duplicated average flights: 0



<div class="alert alert-success">
<h1>Комментарий ревьюера <a class="tocSkip"></a></h1>
Молодец, что не забыл проверить на дубликаты
</div>

### Вывод

Данные корректны, типы соответствуют данным, пропусков и дупликатов нет. Корректировки не нуждаются

## ТОП 10 и построение графиков

### ТОП 10 городов по количеству рейсов

In [8]:
city_top10 = query_3.head(10)
city_top10

city  average_flights
0           Москва       129.774194
1  Санкт-Петербург        31.161290
2      Новосибирск        17.322581
3       Красноярск        11.580645
4     Екатеринбург        11.322581
5   Ростов-на-Дону        10.193548
6            Пермь        10.129032
7           Брянск        10.000000
8             Сочи         9.612903
9        Ульяновск         9.580645

### График ТОП 10 городов по количеству рейсов

In [9]:
fig = px.bar(city_top10.sort_values(by='average_flights',ascending=True),
             y= 'city', x= 'average_flights',text='average_flights', 
             title='ТОП 10 городов и количество рейсов, среднее за день', 
             color_discrete_sequence=px.colors.sequential.RdBu)
fig.update_traces(texttemplate='%{text:.2s}', textposition='outside')
fig.update_layout(uniformtext_minsize=8, uniformtext_mode='hide')
fig.show()


<div class="alert alert-success">
<h1>Комментарий ревьюера <a class="tocSkip"></a></h1>
График топ!
</div>

#### Вывод

Из графика наибольшее количество рейсов Москва и Санк-Петербург, как большие хабы для осуществления полетов.  
Дальше идет группа аэропортов Сибири, Урала и средней полосы, средние хабы  
Группа "южного направления", срез данных сентябрь месяц, "бархатный сезон отпусков"

### График- модели самолётов и количество рейсов

In [10]:

fig = px.pie(query_1, values='flights_amount',
             names='model', title='Модели самолётов и количество рейсов', 
             color_discrete_sequence=px.colors.sequential.RdBu)
fig.update_traces(textposition='inside', textinfo='percent+label')
fig.show()

#### Вывод
Первый три места ханимают следющие модели соответсвено: Cessana 208 Caravan, Bombardier CRJ-20 и Sukhoi Superjet -100.  
По количество полётов первые три места с отрывом отличаются от осталных. Их суммарное порцентное соотношение составляет около 80% от общего количетсво полётов.  
Это местные и региональные самолеты.   
Видно, что у авиакомпании большой объем полетов осуществляется небольшими самолетами


<div class="alert alert-success">
<h1>Комментарий ревьюера <a class="tocSkip"></a></h1>
Молодец, верный анализ
</div>

### График - города и количество рейсов

In [11]:

fig = px.bar(query_3.sort_values(by='average_flights',ascending=True), 
             x='average_flights', y='city', text='average_flights',
             title='Города и количество рейсов, среднее в день',
            color_discrete_sequence=px.colors.sequential.RdBu, height=2000)
fig.update_traces(texttemplate='%{text:.2s}', textposition='outside')
fig.update_layout(uniformtext_minsize=8, uniformtext_mode='hide')
fig.show()

#### Вывод

Из графика видно, что рейсы распределяются на три группы:  
крупные хабы, средние и небольшие аэропорты.

## Вывод

В данном проекте не хватает возможности напрямую из проекта подключиться к pyspark и вытянуть дополнительные аналитики,
т.к. срезы полученные в виде csv не обладают всей информацией.

1. Срез данных за 1 месяц показал, что модели самолетов: *Cessna 208 Caravan*, *Bombardier CRJ-200*, *Sukhoi SuperJet-100*  чаще летали по сравнению с другими фирмами производителями, такими как Boeing и Airbus.  Стоит отметить, что *Cessna 208 Caravan* является лидером по перелетам, но это однотурбинный самолет расчитанный на небольшие расстояния, и используется  для сообщения с небольшими населенными пунктами, где не рентабельно использовать более крупные самолеты и отсутствуют большие аэродромы. 

2. Количество рейсов вероятнее всего соответствует количеству жителей в городе, что  заметно на  графике по соотношению городов и среднему количеству полетов в неделю. 


В дополнение к текущим заданиям, я бы дополнительно исследовал города и топ самолетов для каждого города, сезональные активности, корреляцию между численностью и количеством полетов, модели и вместительность самолетов - пасажиропоток. 


<div class="alert alert-success">
<h1>Комментарий ревьюера v2 <a class="tocSkip"></a></h1>
Классые графки, четкий и лаконичный анализ, лайк!
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"></a></h2>

-Спасибо за качественно сделанный проект, было приятно проверять.

- Код написан хорошо 

- Соблюдена структура проекта 

- Отдельное спасибо за графики



</div>

# Чек-лист проверки

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Файлы изучены
- [x]  Типы данных проверены на корректность
- [x]  Выбран топ-10 городов по количеству рейсов
- [x]  Построен график «Модели самолётов и количество рейсов»
- [x]  Построен график «Города и количество рейсов»
- [x]  Построен график «Топ-10 городов и количество рейсов»
- [x]  На каждом этапе написаны выводы
- [x]  Написан общий вывод